src/core/thread/network_data_tlvs.hpp

```diff
@@ -693,6 +697,250 @@ class CommissioningDataTlv: public NetworkDataTlv
+    };
+
+    /**
+     * This method returns the Sub-TLVs length in bytes.
+     *
+     * @returns The Sub-TLVs length in bytes.
+     *
+     */
+    uint8_t GetSubTlvsLength(void) {
+        return GetLength() - (sizeof(*this) - sizeof(NetworkDataTlv)) - (IsThreadEnterprise() ? 0 : sizeof(uint32_t))
+               - sizeof(uint8_t)/* mServiceDataLength */ - GetServiceDataLength();
+    }
+
+    /**
+     * This method sets the Sub-TLVs length in bytes.
+     *
+     * @param[in]  aLength  The Sub-TLVs length in bytes.
+     *
+     */
+    void SetSubTlvsLength(uint8_t aLength) {
+        SetLength(sizeof(*this) - sizeof(NetworkDataTlv) + (IsThreadEnterprise() ? 0 : sizeof(uint32_t))
+                  + sizeof(uint8_t)/* mServiceDataLength */ + GetServiceDataLength() + aLength);
+    }
+
+    /**
+     * This method returns a pointer to the Sub-TLVs.
+     *
+     * @returns A pointer to the Sub-TLVs.
+     *
+     */
+    NetworkDataTlv *GetSubTlvs(void) {
+        return reinterpret_cast<NetworkDataTlv *>(GetServiceDataLengthLocation() + sizeof(uint8_t) + GetServiceDataLength());
+    }
+
+private:
+    /**
+     * This method returns pointer to where mServiceDataLength would be.
+     *
+     * @returns pointer to service data length location
+     */
+    uint8_t *GetServiceDataLengthLocation(void) {
+        return GetEnterpriseNumberLocation() + (IsThreadEnterprise() ? 0 : sizeof(uint32_t));
+    }
+
+    /**
+     * This method returns pointer to where mEnterpriseNumber would be.
+     *
+     * Note: this method returns uint8_t*, not uint32_t*.
+     *
+     * @returns pointer to enterprise number location
+     */
+    uint8_t *GetEnterpriseNumberLocation(void) {
+        return &mTResSId + sizeof(mTResSId);
+    }
+
+    enum
+    {
+        kTOffset = 7,
+        kTMask = 0x1 << kTOffset,
+        kSIdOffset = 0,
+        kSIdMask = 0xf << kSIdOffset,
+    };
+
+    uint8_t mTResSId;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Server Data TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class ServerTlv: public NetworkDataTlv
+{
+public:
+    /**
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void) { NetworkDataTlv::Init(); SetType(kTypeServer); SetLength(2); }
+
+    /**
+     * This method returns the S_server_16 value.
+     *
+     * @returns The S_server_16 value.
+     */
+    uint16_t GetServer16(void) const { return HostSwap16(mServer16); }
+
+    /**
+     * This method sets the S_server_16 value.
+     *
+     * @param[in]  aServer16  The S_server_16 value.
+     *
+     */
+    void SetServer16(uint16_t aServer16) { mServer16 = HostSwap16(aServer16); }
+
+
+    /**
+     * This method returns a pointer to the Server Data.
+     *
+     * @returns A pointer to the Server Data.
+     */
+    uint8_t *GetServerData(void) { return reinterpret_cast<uint8_t *>(this) + sizeof(*this); }
```

Comment:
Can/should this return a `const` pointer?

Reply:
Well, `const` pointer make no sense when returning, I guess you refer to pointer to `const`. I made this change, though in one place I had to inline this method as a result, minor drawback when compared to the goodness of returning const buffer.